### DISCUSSION TITLE
Typing for @model_validator

### DISCUSSION THREAD

**Discussion by Sanchoyzer** [pydantic/pydantic#8630] [Question]
Let's look at [the example](https://docs.pydantic.dev/latest/concepts/validators/#model-validators) from the Pydantic documentation
```
from typing import Any

from pydantic import BaseModel, ValidationError, model_validator


class UserModel(BaseModel):
    username: str
    password1: str
    password2: str

    @model_validator(mode='before')
    @classmethod
    def check_card_number_omitted(cls, data: Any) -> Any:
        if isinstance(data, dict):
            assert (
                'card_number' not in data
            ), 'card_number should not be included'
        return data
```
1. Why is `data` of type `Any` instead of `Mapping`/ `dict` / `Mapping[str, str]` / `dict[str, str]`?
2. Why does `if isinstance(data, dict)` exist?

**Comment by salomon-smekecohen** [comment]
Hey @Sanchoyzer , I am not a maintainer but I believe I can answer this.

That model_validator is of mode "before", meaning it happens before the attributes have been coerced/type checked.

`UserModel(username=object())` would cause an object to be passed to check_card_number_omitted (hence type is `Any`) and then you would want the instance check so that your `in` test is valid within the function. Mypy would not allow membership testing an Any, and it could potentially raise attribute or type errors.

**Reply by Sanchoyzer** [reply]
Thanks
